django/thunderstore/repository/models/comment.py

```diff
@@ -0,0 +1,65 @@
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models import signals
+from django.dispatch import receiver
+from ulid2 import generate_ulid_as_uuid
+
+from thunderstore.core.mixins import TimestampMixin
+
+
+class Comment(TimestampMixin, models.Model):
+    thread = GenericForeignKey("thread_content_type", "thread_object_id")
+    thread_content_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.CASCADE,
+        related_name="comments",
+    )
+    thread_object_id = models.PositiveIntegerField()
+    parent_comment = models.ForeignKey(
+        "self",
+        on_delete=models.CASCADE,
+        related_name="replies",
+        null=True,
+    )
+    # author should never be NULL
+    # It is temporarily set before the `User` `pre_delete` signal is ran but then set
+    # to a "ghost user" `User`.
+    author = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        null=True,
+        on_delete=models.SET_NULL,
+        related_name="comments",
+    )
+    content = models.TextField(
+        max_length=2048,
+    )
+    uuid4 = models.UUIDField(
+        default=generate_ulid_as_uuid,
+        editable=False,
+        unique=True,
+        primary_key=True,
+    )
+    is_pinned = models.BooleanField(
+        default=False,
+    )
+
+
+def _create_ghost_user_username(id_: str) -> str:
+    # The ID used in ghost user usernames and emails are not the same as their `User.id`
+    # This is because `User` still uses an autoincrement ID
+    return f"{id_}.gu@thunderstore.io"
+
+
+@receiver(signals.pre_delete, sender=settings.AUTH_USER_MODEL)
+def set_ghost_user(instance, **kwargs) -> None:
+    if not instance.comments.exists():
+        return
+    uuid = generate_ulid_as_uuid()
+    username = _create_ghost_user_username(uuid.hex)
+    ghost_user = User.objects.create_user(username, email=username)
```

Comment:
If we display the username on the frontend on comments, showing this email is probably fairly ugly. Maybe setting one of the following could help:
1. `first_name` to `Deleted User`
2. Setting `username` to `uuid.hex`

Reply:
I don't really like the idea of using usernames for display names to begin with, IMO they should be used as user identifiers and nothing more. To that end I feel like the best way to handle these users is akin to how we handle services accounts, which is to generate an email address under our domain to be used as their username + email.

For display purposes we should come up with something else, but the format `Deleted User {some id}` sounds good

Reply:
Where will we actually get what to display from? Their `first_name` and `last_name`? Because we don't have our own user model we can't make our own property so standardising how we display users might be difficult, even with a util function.

Reply:
first + last name would be good if we're using the Django user model fields, but we can't really use those since they might inadvertedly reveal people's names (e.g. GitHub oauth could've returned their github profile name, which gets mapped to the django model), so in favor of privacy by default, let's not do that.

Having a related model (akin to service accounts) with extra user metadata would be the best way to extend the user model, but probably a bit overkill to do just yet.

So what we should do now is probably to:
- Set username according the the display name `Deleted User {id}`
- Set email according to a generated email akin to how service account emails are generated
- Later on when we have moved on from usernames as display names, we can migrate the emails to be usernames. This is probably a long way off and could see a larger user-model refactor.

So as far as this PR goes, this should be good if we just autogenerate a service-account-esque email for them.

Reply:
Another minor nitpick: Maybe we should refer to deleted users as `deleted user` internally as well, just to keep the language consistent with the frontend. Or was there a specific reason for picking something else?

Reply:
I just picked `ghost user` just so it was clear that I wasn't just talked about a user who is deleted but instead a special `User` type (based on GitHub's @ghost). I think being able to differentiate these is important, especially when we can't refer to a specific property in comments and internal docs but the name isn't neccesarily special.

Reply:
In that case, maybe we should have the frontend-facing user also be called `Ghost User`?

Reply:
I don't have a strong preference but I'm not sure if `Ghost User` is clear enough?

Reply:
I'd just like to keep the naming consistent between internal vs. external. GitHub has Ghost User as the public facing one so I think it's fine, we can have some description on the profile page if/when we have profile pages